CMakeLists.txt

```diff
@@ -1,504 +1,2025 @@
-#
-# Don't put any code above or in this block, unless it
-# has similar constraints.
-cmake_minimum_required(VERSION 3.1.0)
-set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/Builds/CMake")
-include(CMakeFuncs)
-set(openssl_min 1.0.2)
-parse_target()
-project(rippled)
-#########################################################
-
-if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
-  set(dir "build")
-  set(cmd "cmake")
-  if (target)
-    set(dir "${dir}/${target}")
-    set(cmd "${cmd} -Dtarget=${target}")
-  elseif(CMAKE_BUILD_TYPE)
-    set(dir "${dir}/${CMAKE_BUILD_TYPE}")
-    set(cmd "${cmd} -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}")
-  else()
-    set(dir "${dir}/default")
-  endif()
-  set(cmd "${cmd} ${CMAKE_SOURCE_DIR}")
-
-  message(FATAL_ERROR "Builds are not allowed in ${CMAKE_SOURCE_DIR}.\n"
-    "Instead:\n"
-    "1) Remove the CMakeCache.txt file and CMakeFiles directory "
-   "from ${CMAKE_SOURCE_DIR}.\n"
-    "2) Create a directory to hold your build files, for example: ${dir}.\n"
-    "3) Change to that directory.\n"
-    "4) Run cmake targetting ${CMAKE_SOURCE_DIR}, for example: ${cmd}")
-endif()
-if("${CMAKE_GENERATOR}" MATCHES "Visual Studio" AND
-    NOT ("${CMAKE_GENERATOR}" MATCHES .*Win64.*))
-  message(FATAL_ERROR "Visual Studio 32-bit build is unsupported. Use
-    -G\"${CMAKE_GENERATOR} Win64\"")
-endif()
-
-setup_build_cache()
-
-if(nonunity)
-    get_cmake_property(allvars VARIABLES)
-    string(REGEX MATCHALL "[^;]*(DEBUG|RELEASE)[^;]*" matchvars "${allvars}")
-    foreach(var IN LISTS matchvars)
-        string(REGEX REPLACE "(DEBUG|RELEASE)" "\\1CLASSIC" newvar ${var})
-        set(${newvar} ${${var}})
-    endforeach()
-
-    get_cmake_property(allvars CACHE_VARIABLES)
-    string(REGEX MATCHALL "[^;]*(DEBUG|RELEASE)[^;]*" matchvars "${allvars}")
-    foreach(var IN LISTS matchvars)
-        string(REGEX REPLACE "(DEBUG|RELEASE)" "\\1CLASSIC" newvar ${var})
-        set(${newvar} ${${var}} CACHE STRING "Copied from ${var}")
-    endforeach()
-endif()
-
-determine_build_type()
-
-check_gcc4_abi()
-
-############################################################
-
-include_directories(
-  src
-  src/beast
-  src/beast/extras
-  src/nudb/include
-  src/soci/src
-  src/soci/include)
-
-special_build_flags()
-
-############################################################
-
-use_boost(
-  # resist the temptation to alphabetize these. coroutine
-  # must come before context.
-  chrono
-  coroutine
-  context
-  date_time
-  filesystem
-  program_options
-  regex
-  serialization
-  system
-  thread)
-
-use_pthread()
-
-use_openssl(${openssl_min})
-
-use_protobuf()
 
-setup_build_boilerplate()
+cmake_minimum_required (VERSION 3.9.0)
+set (CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/Builds/CMake ${CMAKE_MODULE_PATH})
+include (CMakeFuncs)
+include (CheckCXXCompilerFlag)
+if (target)
+  message (WARNING
+    "The target option is deprecated and will be removed in a future release")
+  parse_target()
```

Comment:
Removing `target` pretty much requires calling `cmake` from a script for anything other than a default build. Compare:
```
  cmake -Dtarget=gcc.debug.unity
  cmake -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_BUILD_TYPE=Debug -Dunity=ON ..
 ```
No one is going to type the second command more than once before writing a script. So we do simplify the build script by removing this, but at the cost of pushing the complexity onto other developers.

No change needed, but we should address this when we do remove `target`

Reply:
just want to note that alternative to what you've written:

```
CC=gcc CXX=g++ cmake -DCMAKE_BUILD_TYPE=Debug
```
(since unity defaults to on, you don't have to specify it). Also, we could choose to make the build type default to Debug since the "unspecified" build type is cmake is pretty useless except for validating that compiling works)

Reply:
I'm in favor of defaulting the `CMAKE_BUILD_TYPE` to Debug. Also, one thing to keep in mind is how a lot of these variables you're passing to `configure` at CMake time will be in your `CMakeCache`. I rarely need to reconfigure, and even if I do, I am typically using the same compiler. So, it suffices to just have an out of source build folder, configure once, and then invoke Ninja many times without rerunning CMake.

Reply:
I would rather default to "release". There are so extremely few external developers compared to people running the software that it seems like the most common use case is to build a release build.

Reply:
I agree with @MarkusTeufelberger I'd rather it default to `Release`

Reply:
Fair point @MarkusTeufelberger